internal/provider/resource_access_generic_schema.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package provider
+
+import (
+	jimmnames "github.com/canonical/jimm-go-sdk/v3/names"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/setvalidator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/juju/names/v4"
+)
+
+type genericSchema map[string]schema.Attribute
+
+// baseAccessSchema returns a map of schema attributes for a JAAS access resource.
+// Access resources should use this schema and add any additional attributes e.g. name or uuid.
+func baseAccessSchema() genericSchema {
+	return map[string]schema.Attribute{
+		"access": schema.StringAttribute{
+			Description: "Level of access to grant. Changing this value will replace the Terraform resource. Valid access levels are described at https://canonical-jaas-documentation.readthedocs-hosted.com/en/latest/reference/authorisation_model/#valid-relations",
+			Required:    true,
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.RequiresReplace(),
+			},
+		},
+		"users": schema.SetAttribute{
+			Description: "List of users to grant access.",
+			Optional:    true,
+			ElementType: types.StringType,
+			Validators: []validator.Set{
+				setvalidator.ValueStringsAre(ValidatorMatchString(names.IsValidUser, "email must be a valid Juju username")),
+				setvalidator.ValueStringsAre(stringvalidator.RegexMatches(basicEmailValidationRe, "email must contain an @ symbol")),
+			},
+		},
+		"groups": schema.SetAttribute{
+			Description: "List of groups to grant access.",
+			Optional:    true,
+			ElementType: types.StringType,
+			Validators: []validator.Set{
+				setvalidator.ValueStringsAre(ValidatorMatchString(jimmnames.IsValidGroupId, "group ID must be valid")),
+			},
+		},
+		"service_accounts": schema.SetAttribute{
+			Description: "List of service accounts to grant access.",
```

Comment:
doc: the description will be user documentation. what string is considered a valid service account?